client/network/src/gossip.rs

```diff
@@ -0,0 +1,318 @@
+	///
+	/// See [`DirectGossip::new`] for details.
+	pub fn build<M, F>(
+		self,
+		protocol: ConsensusEngineId,
+		queue_size_limit: usize,
+		messages_encode: F
+	) -> (DirectedGossip<M>, impl Future<Output = ()> + Send + 'static)
+	where
+		M: Send + 'static,
+		F: Fn(M) -> Vec<u8> + Send + 'static,
+	{
+		let shared = Arc::new(Shared {
+			stop_task: atomic::AtomicBool::new(false),
+			condvar: Condvar::new(),
+			queue_size_limit,
+			locked: Mutex::new(SharedLock {
+				messages_queue: VecDeque::with_capacity(queue_size_limit),
+			}),
+		});
+
+		let task = spawn_task(
+			self.service,
+			self.peer_id,
+			protocol,
+			shared.clone(),
+			messages_encode
+		);
+
+		(DirectedGossip { shared }, task)
+	}
+}
+
+impl fmt::Debug for DirectedGossipPrototype {
+	fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+		f.debug_struct("DirectedGossipPrototype")
+			.field("peer_id", &self.peer_id)
+			.finish()
+	}
+}
+
+/// Locked queue of messages to the given peer.
+///
+/// As long as this struct exists, the background task is asleep and the owner of the [`QueueLock`]
+/// is in total control of the buffer.
+#[must_use]
+pub struct QueueLock<'a, M> {
+	lock: MutexGuard<'a, SharedLock<M>>,
+	condvar: &'a Condvar,
+	/// Same as [`Shared::queue_size_limit`].
+	queue_size_limit: usize,
+}
+
+impl<'a, M: Send + 'static> QueueLock<'a, M> {
+	/// Pushes a message to the queue, or discards it if the queue is full.
+	pub fn push_or_discard(&mut self, message: M) {
+		if self.lock.messages_queue.len() < self.queue_size_limit {
+			self.lock.messages_queue.push_back(message);
+		}
+	}
+
+	/// Pushes a message to the queue. Does not enforce any limit to the size of the queue.
+	/// Use this method only if the message is extremely important.
+	pub fn push_unbounded(&mut self, message: M) {
+		self.lock.messages_queue.push_back(message);
+	}
+
+	/// Calls `filter` for each message in the queue, and removes the ones for which `false` is
+	/// returned.
+	///
+	/// > **Note**: The parameter of `filter` is a `&M` and not a `&mut M` (which would be
+	/// >           better) because the underlying implementation relies on `VecDeque::retain`.
+	pub fn retain(&mut self, filter: impl FnMut(&M) -> bool) {
+		self.lock.messages_queue.retain(filter);
+	}
+}
+
+impl<'a, M> Drop for QueueLock<'a, M> {
+	fn drop(&mut self) {
+		// We notify the `Condvar` in the destructor in order to be able to push multiple
+		// messages and wake up the background task only one afterwards.
+		self.condvar.notify_one();
+	}
+}
+
+#[derive(Debug)]
+struct Shared<M> {
+	/// Read by the background task after locking `locked`. If true, the task stops.
+	stop_task: atomic::AtomicBool,
+	locked: Mutex<SharedLock<M>>,
+	/// Must be notified every time the content of `locked` changes.
+	condvar: Condvar,
+	/// Maximum number of elements in `messages_queue`.
+	queue_size_limit: usize,
+}
+
+#[derive(Debug)]
+struct SharedLock<M> {
```

Comment:
Unless you plan to add more to `SharedLock` I suggest to remove the type and use the `Mutex` directly around the `VecDeque`. If you want to keep it I think a name such as `Queue` might be more apt.